### PR TITLE
Add fork type clarifications for light client headers while constructing light client data

### PR DESCRIPTION
**Pull request by g11tech**
Since there could be some confusion wile constructing the light client data especially at fork boundaries, this PR attempts to add some clarifications which were discussed in the R&D discord light client channel with @etan-status and others.

### CODE REVIEW

File: specs/capella/light-client/full-node.md [hunk 1]
```diff
@@ -76,3 +76,31 @@ def block_to_light_client_header(block: SignedBeaconBlock) -> LightClientHeader:
         execution_branch=execution_branch,
     )
 ```
+
+## Deriving light client data
+
+### `create_light_client_update`
+
+Full nodes SHOULD continue deriving the `LightClientUpdate` for each sync committee period as per the altair specs but with the modified or upgraded `attested_header` and `finalized_header` with the following considerations:
```
Comment (etan-status) [inline specs/capella/light-client/full-node.md:84]: Hmm. The way how `consensus-specs` work is that the later forks are overlays. So, the phase0 spec is used as basis, then altair is overlayed on it (adding new functions, and replacing redefined ones), then bellatrix is overlayed, then capella. 

So, technically, the "altair specs" being referred here are also copied to capella as is, with `block_to_light_client_header` getting overridden.

That's sort of what the "introduction" above mentions:
> It extends the [Altair Light Client specifications](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md).

File: specs/capella/light-client/full-node.md [hunk 2]
```diff
@@ -76,3 +76,31 @@ def block_to_light_client_header(block: SignedBeaconBlock) -> LightClientHeader:
...
+
+- Both `attested_header` and `finalized_header` are of the same fork types as `LightClientUpdate` and might need upgradation to bundle as update.
```
Comment (etan-status) [inline specs/capella/light-client/full-node.md:86]: "upgradation" > "upgrading"?

Also, the `capella.block_to_light_client_header` function above works perfectly fine on altair blocks as well. So, whether an implementation first gets the original representation and upgrades from there, or whether it immediately computes the correct representation doesn't matter that much.

File: specs/capella/light-client/full-node.md [hunk 3]
```diff
@@ -76,3 +76,31 @@ def block_to_light_client_header(block: SignedBeaconBlock) -> LightClientHeader:
...
+- Above implies that the first update of capella MIGHT have both `attested_header` and `finalized_header` upgraded (if capella hardfork coincides with a new period) as they might belong to pre-capella epochs. Post this `finalized_header` will need upgradation till a new finalization occurs in capella.
```
Comment (etan-status) [inline specs/capella/light-client/full-node.md:87]: The `attested_header` determines the spec fork to use as basis for building the data.
This is mentioned in the Altair [p2p-interface.md](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/p2p-interface.md).

> The gossip ForkDigest-context is determined based on `compute_fork_version(compute_epoch_at_slot(optimistic_update.attested_header.beacon.slot))`.

The entire object will be built using the logic from that spec fork. 

The first case cannot really happen for data production, the `attested_header` will never need an upgrade, as it is the one defining the spec fork. Spec fork is not determined by wall clock or `signature_slot`.

Also, the "first update" is not the only one that's special, depending on branching situation and missed blocks, you could even set up hypothetical scenarios where `finalized` is altair, `attested` is capella, and `signature_slot` is deneb. the unit tests construct some of those cases to cover edge cases

File: specs/capella/light-client/full-node.md [hunk 4]
```diff
@@ -76,3 +76,31 @@ def block_to_light_client_header(block: SignedBeaconBlock) -> LightClientHeader:
...
+- Over the wire, the fork information of transmitted data might be available (like fork `version` in beacon-apis or fork `context` in req/resp or fork gossip `topics`)
```
Comment (etan-status) [inline specs/capella/light-client/full-node.md:88]: It is always available, otherwise the protocol doesn't work.

libp2p req/resp uses the `ForkDigest`-context: https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/p2p-interface.md#forkdigest-context

libp2p gossip has the `ForkDigestValue` in the topic name: https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#topics-and-messages

REST uses a mix of Eth-Consensus-Version HTTP header, binary interleaved ForkDigest, or JSON "version" fields to indicate spec fork.

File: specs/capella/light-client/full-node.md [hunk 5]
```diff
@@ -76,3 +76,31 @@ def block_to_light_client_header(block: SignedBeaconBlock) -> LightClientHeader:
...
+- However `LightClientUpdate`'s base fork can be assumed belonging to `attested_header`'s slot and can be used to store data without loss of information.
```
Comment (etan-status) [inline specs/capella/light-client/full-node.md:89]: `LightClientStore` can actually be from the future. To process data from older forks against newer `LightClientStore`, that's the only time where you need the [upgrade functions](https://github.com/ethereum/consensus-specs/blob/dev/specs/capella/light-client/fork.md).

So, you could ship a Deneb `LightClientStore` today and still process Altair/Capella data with it. Wouldn't recommend doing so, as Deneb is still not a final spec, but it is possible, e.g., on devnets and so on.

**Review by etan-status — COMMENTED** [review]
I think the goal of adding more clarifying comments is nice here.

However, @hwwhww, it would be really awesome if there was a viewer for these specs that shows the effective spec with all the overlays applied. Especially in the light client part, where most logic gets inherited from Altair, the current approach is quite confusing, and was mentioned a couple times, e.g., @ Capella interop, and also in #light-clients 

Maybe some crazy tooling similar to how it exists for the EIPs, or for the beacon APIs, that creates a nice webpage out of the GitHub documents.

I know it's not trivial to do, especially with non-code overlays, such as the gossip tables in p2p-interface.md, but you are probably the closest to know the needs and pitfalls for such a viewer.
